### ISSUE TITLE
Support for customizing JS chunk granularity when using deferred loading in Flutter for Web

### ISSUE THREAD

**Issue by zjrwdl** [flutter/flutter#127859]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

Currently, when using deferred loading in Flutter for Web, the Dart compiler generates multiple JS chunks for each deferred library. In some cases, this can lead to a large number of small JS chunks, which may negatively impact the performance of the web application. It would be beneficial for developers to have the ability to customize the granularity of the generated JS chunks, allowing them to control the number and size of the chunks according to their specific needs.

Steps:
1.Create a new Flutter for Web project.
2.Add multiple deferred libraries to the project, each containing a small amount of code.
3.Build the project using flutter build web.
4.Observe that the Dart compiler generates a separate JS chunk for each deferred library, potentially leading to a large number of small JS chunks.

deferredLibraryParts in main.dart.js contains js parts of every deferred module.
home_page:[0,1,2,3,4,28,20,5,40,29,21,6,50,30,22,7,41,42,31,51,43,32,8,67,68,52,53,44,77,71,54,55,33,78,72,69,56,57,58,45,86,80,73,74,59,60,35,87,88,81,75,61,47,90,94],
in our project , home_page contains dozens of js file.

### Proposal

Most modern browsers allow 6 to 8 simultaneous connections per domain name. This means that if your site has too many JS files, it may take multiple roundtrips for the browser to download all the files. This increases page load time, which affects user experience.

It would be helpful if the Dart compiler provided an option for developers to customize the granularity of the generated JS chunks when using deferred loading. This could be achieved by allowing developers to specify a threshold for the size of the chunks, or by providing a way to manually group deferred libraries into larger chunks. This feature would give developers more control over the performance of their web applications and help them optimize the loading process according to their specific requirements.

Please consider adding support for customizing JS chunk granularity when using deferred loading in Flutter for Web. This would be a valuable addition to the framework and would greatly benefit developers working on web applications with complex loading requirements.

**Comment by danagbemava-nc** [comment]
Hi @zjrwdl, thanks for filing this proposal. Could you provide a small code sample in a repo that further illustrates the issue you're referring to in your steps section? 

Thank you

**Comment by zjrwdl** [comment]
> Thanks for the info, I can reproduce it with the gallery app.
> 
> Labeling as a proposal for further insight from the team

Thanks, looking forward to your reply

**Comment by kevmoo** [comment]
@sigmundch should comment here...

**Comment by sigmundch** [comment]
Good news is that there is a lot available today to help here, the bad news is that it's not well documented. Also, the APIs here still needs improvements which are not currently prioritized.

Here are a couple things worth sharing:
* The `.part` files today are designed to support bundling by simply concatenating files. Dart2js offers a hook that allows developers to intercept deferred-load requests at runtime. This has been used successfully by customers to reinterpret the meaning of a request and dynamically bundle together multiple requests in one go. In particular:
  * the compiler has a flag `--deferred-map` that will give you a .json file with all the data needed to understand load lists and which files are needed on a given load (documentation is pretty minimal but you can read it [here](https://github.com/dart-lang/sdk/blob/70064fcdbb2d0f5a40a02274f74a9fbfdc5c292e/pkg/compiler/lib/src/dart2js.dart#L1173) or running `dart compile js -v -h`)
  * the compiler also allows you to define a top-level JavaScript function called [`dartDeferredLibraryLoader`](https://github.com/dart-lang/sdk/blob/70064fcdbb2d0f5a40a02274f74a9fbfdc5c292e/pkg/compiler/lib/src/js_emitter/headers.dart#L30) that will override the loading logic. This function can then request not one individual file, but a bundle of concatenated files as a single request based on the load-id (and some tracking to share what has already been loaded). A server that has access to the deferred-map can then dynamically concatenate files, send them to the client, and the client hook can then satisfy all requests at once.

* Separately, dart2js also offers a compile-time hook that allows developers to specify [constraints](https://github.com/dart-lang/sdk/blob/70064fcdbb2d0f5a40a02274f74a9fbfdc5c292e/pkg/compiler/lib/src/deferred_load/deferred_load.dart#L221) that help the compiler to merge part files together in advance. For example, if the compiler sees 2 deferred imports, it may create 3 part files: one for the common logic between the two, and one for the logic unique to each import. This helps accommodate for loading the imports in any order. The compile-time hook allows developers to specify constraints, like saying that one import is always going to be loaded before another. When that is available, some parts can be merged together. Some internal customers have used this and were able to dramatically cut down the number of files loaded.

The tradeoffs between the two approaches are interesting, and we've seen internal customers choose differently here based on their priorities. The compile-time hook requires tuning and monitoring over time (since the application will change and new constraints need to be added), while the dynamic approach is more flexible and easy to set up once and forget about it. That said, the compile-time hook doesn't require any client or server-side logic to make it work, while the dynamic approach does.

All of that said, I have little pointers to share for how to use these, unfortuantely. I filed a couple issues to follow up on the Dart side on documentation and feature improvements:
* https://github.com/dart-lang/sdk/issues/52660
* https://github.com/dart-lang/sdk/issues/51837

**Comment by zjrwdl** [comment]
> * dart2js: write documentation for defer loading hooks dart-lang/sdk#52660

Thank you for your support, I will learn first and look forward to your more progress

**Comment by zjrwdl** [comment]
> Good news is that there is a lot available today to help here, the bad news is that it's not well documented. Also, the APIs here still needs improvements which are not currently prioritized.
> 
> Here are a couple things worth sharing:
> 
> * The `.part` files today are designed to support bundling by simply concatenating files. Dart2js offers a hook that allows developers to intercept deferred-load requests at runtime. This has been used successfully by customers to reinterpret the meaning of a request and dynamically bundle together multiple requests in one go. In particular:
>   
>   * the compiler has a flag `--deferred-map` that will give you a .json file with all the data needed to understand load lists and which files are needed on a given load (documentation is pretty minimal but you can read it [here](https://github.com/dart-lang/sdk/blob/70064fcdbb2d0f5a40a02274f74a9fbfdc5c292e/pkg/compiler/lib/src/dart2js.dart#L1173) or running `dart compile js -v -h`)
>   * the compiler also allows you to define a top-level JavaScript function called [`dartDeferredLibraryLoader`](https://github.com/dart-lang/sdk/blob/70064fcdbb2d0f5a40a02274f74a9fbfdc5c292e/pkg/compiler/lib/src/js_emitter/headers.dart#L30) that will override the loading logic. This function can then request not one individual file, but a bundle of concatenated files as a single request based on the load-id (and some tracking to share what has already been loaded). A server that has access to the deferred-map can then dynamically concatenate files, send them to the client, and the client hook can then satisfy all requests at once.
> * Separately, dart2js also offers a compile-time hook that allows developers to specify [constraints](https://github.com/dart-lang/sdk/blob/70064fcdbb2d0f5a40a02274f74a9fbfdc5c292e/pkg/compiler/lib/src/deferred_load/deferred_load.dart#L221) that help the compiler to merge part files together in advance. For example, if the compiler sees 2 deferred imports, it may create 3 part files: one for the common logic between the two, and one for the logic unique to each import. This helps accommodate for loading the imports in any order. The compile-time hook allows developers to specify constraints, like saying that one import is always going to be loaded before another. When that is available, some parts can be merged together. Some internal customers have used this and were able to dramatically cut down the number of files loaded.
> 
> The tradeoffs between the two approaches are interesting, and we've seen internal customers choose differently here based on their priorities. The compile-time hook requires tuning and monitoring over time (since the application will change and new constraints need to be added), while the dynamic approach is more flexible and easy to set up once and forget about it. That said, the compile-time hook doesn't require any client or server-side logic to make it work, while the dynamic approach does.
> 
> All of that said, I have little pointers to share for how to use these, unfortuantely. I filed a couple issues to follow up on the Dart side on documentation and feature improvements:
> 
> * [dart2js: write documentation for defer loading hooks dart-lang/sdk#52660](https://github.com/dart-lang/sdk/issues/52660)
> * [proposal: parameterize deferred loading runtime dart-lang/sdk#51837](https://github.com/dart-lang/sdk/issues/51837)

After studying the above information, I have the following insights and suggestions:
1. The compile-time hook with constraint cannot solve the scenario of loading the first page (assuming homepage: [0,1,2,...100] contains many parts, many of which belong to <1kb debris)
2. Dynamic approach Theoretically, developers can manually implement splicing of fragments before sending them out, but it is not suitable for some scenarios. For example, the flutter web construction product is deployed on a COS (Cloud Object Storage) cloud server, and it is inconvenient for developers to customize the splicing logic. . In short, for developers, there are some adaptation costs. So I personally think it is not the best solution.
3. In the problem description, I mentioned that the dart deferred algorithm will generate a lot of part fragments. For a medium and large website, the number of part fragments may be very large. It may appear that loading a homepage requires loading hundreds of part.js files, which brings performance problems to loading web pages. So strictly speaking, this is a bug of flutter web. It is not a good way for every developer to customize and modify it. I think this problem is best handled on the compilation side, such as:
(1) The deferred algorithm supports configuration and provides option switching on the flutter build web side. The algorithm includes the current default, and then add other possible solutions. For example, our current desired effect is: assuming 10 deferred imports, the algorithm generated products are: main.dart.js, main.dart.part1.js, main.dart.part2 .js, ...main.dart.part10.js.
page1:0,1
page2:0,2
...
page10:0,10
In short, the current deferred algorithm fragmentation problem is a common problem for flutter web developers, and I hope it can be supported and further solved in the official version.
Thanks

**Comment by yjbanov** [comment]
If this is still an issue, consider filing an issue at https://github.com/dart-lang/sdk/issues

**Comment by sigmundch** [comment]
@zjrwdl - we really appreciate all your feedback and thanks for digging into this in detail.

One thing you mentioned that caught me by surprise is that you said that you are seeing many part files on the first load. The idea behind deferred loading is to defer code that is not needed upfront. If the code is needed while the app is being loaded, ideally it should not be deferred in the first place. I'm curious what is causing this effect for you:
* If it is accidental and you own that portion of the code, this could be addressed by removing a few deferred imports
* If you don't own the code with the deferred imports, this could be addressed with program-split-constraints (where you specify at compile time how to combine pieecs) or disabling deferred loading with `--no-program-split (equivalent to a split constraint that merges everything).
* If the reason is that your app has multiple landing pages, then a dynamic approach would be more suitable. But as you said, the configuration for this is hard.

Aside from the issues I mentioned earlier, I wanted to share we also have another issue to track improving the deferred loading part files. Especially when we can reduce the number of small files that provide little value (see https://github.com/dart-lang/sdk/issues/52870).

Also worth sharing that in the past we tried making automated bundling that emitted only a certain number of files (e.g. you specify a flag to put a limit in the number of files), but this proved to be very hard to implement in the compiler in a way that yielded good performance results.

@yjbanov - I think there is one item to follow up on the Flutter side, which is to provide more guidance here. While we can add some level of documentation for how to customize the compiler to improve things a bit, I think the bigger challenge here is how to approach the serving stack. Do we have good documentation and guidance for customers on this? If not, should we have some?
